src/SimpleRestServer.ts

```diff
@@ -32,92 +40,74 @@ export class BaseServer<RequestType, ResponseType> {
     }
 
-    getContext(context: NormalizedRequest): FakeRestContext {
+    getContext(normalizedRequest: NormalizedRequest): FakeRestContext {
         for (const name of this.database.getSingleNames()) {
-            const matches = context.url?.match(
+            const matches = normalizedRequest.url?.match(
                 new RegExp(`^${this.baseUrl}\\/(${name})(\\/?.*)?$`),
             );
             if (!matches) continue;
             return {
-                ...context,
+                ...normalizedRequest,
                 single: name,
             };
         }
 
-        const matches = context.url?.match(
+        const matches = normalizedRequest.url?.match(
             new RegExp(`^${this.baseUrl}\\/([^\\/?]+)(\\/(\\w))?(\\?.*)?$`),
         );
         if (matches) {
             const name = matches[1];
             const params = Object.assign(
                 {},
                 this.defaultQuery(name),
-                context.params,
+                normalizedRequest.params,
             );
 
             return {
-                ...context,
+                ...normalizedRequest,
                 collection: name,
                 params,
             };
         }
 
-        return context;
-    }
-
-    getNormalizedRequest(request: RequestType): Promise<NormalizedRequest> {
-        throw new Error('Not implemented');
-    }
-
-    respond(
-        response: BaseResponse | null,
-        request: RequestType,
-        context: FakeRestContext,
-    ): Promise<ResponseType> {
-        throw new Error('Not implemented');
+        return normalizedRequest;
     }
 
-    async handle(request: RequestType): Promise<ResponseType | undefined> {
-        const context = this.getContext(
-            await this.getNormalizedRequest(request),
-        );
-
+    async handle(normalizedRequest: NormalizedRequest): Promise<BaseResponse> {
+        const context = this.getContext(normalizedRequest);
         // Call middlewares
         let index = 0;
         const middlewares = [...this.middlewares];
 
-        const next = (req: RequestType, ctx: FakeRestContext) => {
+        const next = (context: FakeRestContext) => {
             const middleware = middlewares[index++];
             if (middleware) {
-                return middleware(req, ctx, next);
+                return middleware(context, next);
             }
-
-            return this.handleRequest(req, ctx);
+            return this.handleRequest(context);
         };
 
         try {
-            const response = await next(request, context);
-            if (response != null) {
-                return this.respond(response, request, context);
-            }
+            const response = await next(context);
+            return response;
         } catch (error) {
             if (error instanceof Error) {
                 throw error;
             }
 
-            return error as ResponseType;
+            return error as BaseResponse;
```

Comment:
This forbids us to throw a real response (of the type expected by the mock implementation). I would move the try/catch logic in the adapters